patches/server/0731-Add-critical-damage-API.patch

```diff
@@ -0,0 +1,167 @@
++    private boolean critical;
++    public boolean isCritcal() {
++        return this.critical;
++    }
++    public DamageSource critical() {
++        return critical(true);
++    }
++    public DamageSource critical(boolean critical) {
++        this.critical = critical;
++        return this;
++    }
++    // Paper end
+ 
+     public static DamageSource sting(LivingEntity attacker) {
+         return new EntityDamageSource("sting", attacker);
+diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
+index 19980b2d627eb3cacf8d0c3e6785ad2206910fbc..dcc3850fcdcf27a7f5c367a9fc1fe11be448fe5e 100644
+--- a/src/main/java/net/minecraft/world/entity/player/Player.java
++++ b/src/main/java/net/minecraft/world/entity/player/Player.java
+@@ -1251,7 +1251,7 @@ public abstract class Player extends LivingEntity {
+                     }
+ 
+                     Vec3 vec3d = target.getDeltaMovement();
+-                    boolean flag5 = target.hurt(DamageSource.playerAttack(this), f);
++                    boolean flag5 = target.hurt(DamageSource.playerAttack(this).critical(flag2), f); // Paper - add critical damage API
+ 
+                     if (flag5) {
+                         if (i > 0) {
+@@ -1279,7 +1279,7 @@ public abstract class Player extends LivingEntity {
+ 
+                                 if (entityliving != this && entityliving != target && !this.isAlliedTo(entityliving) && (!(entityliving instanceof ArmorStand) || !((ArmorStand) entityliving).isMarker()) && this.distanceToSqr((Entity) entityliving) < 9.0D) {
+                                     // CraftBukkit start - Only apply knockback if the damage hits
+-                                    if (entityliving.hurt(DamageSource.playerAttack(this).sweep(), f4)) {
++                                    if (entityliving.hurt(DamageSource.playerAttack(this).sweep().critical(flag2), f4)) { // Paper - add critical damage API
+                                     entityliving.knockback(0.4000000059604645D, (double) Mth.sin(this.getYRot() * 0.017453292F), (double) (-Mth.cos(this.getYRot() * 0.017453292F)), this); // Paper
+                                     }
+                                     // CraftBukkit end
+diff --git a/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java b/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
+index 6636845ea044c3810e1880aad8b679134cd33668..1d05ad1279d9b1cdd03a8198d58236eec0f34f89 100644
+--- a/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
++++ b/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
+@@ -380,7 +380,7 @@ public abstract class AbstractArrow extends Projectile {
+                 ((LivingEntity) entity1).setLastHurtMob(entity);
+             }
+         }
+-
++        if(this.isCritArrow()) damagesource = damagesource.critical(); // Paper - add critical damage API
```

Comment:
Try to not replace lines when possible, so just make it add a newline, instead of replace a blank like with your addition. There are several other places where you do this, like when you add `@ Deprecated` annotations and whatnot.